INTERNAL ERROR with `__hash__ = None`

<!--
  Use this form only if mypy reports an "INTERNAL ERROR" and/or gives a traceback.
  Please include the traceback and all other messages below (use `mypy --show-traceback`).
-->

**Crash Report**

Mypy crashed when run on the following reduced file. Setting `__hash__` to `None` is the way to suppress Python's default hash method.

**Traceback**

```
> mypy mypy_crash.py --show-traceback
mypy_crash.py:2: error: Incompatible types in assignment (expression has type "None", base class "object" defined the type as "Callable[[object], int]")
mypy_crash.py:5: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.820+dev.15bd486a67a3214fdb854ba2f1d6d524113da6c1
Traceback (most recent call last):
  File "/home/user/.virtualenvs/mypy/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/__main__.py", line 11, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/main.py", line 98, in main
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 180, in build
    sources, options, alt_lib_path, flush_errors, fscache, stdout, stderr, extra_plugins
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 253, in _build
    graph = dispatch(sources, manager, stdout)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 2688, in dispatch
    process_graph(graph, manager)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 3012, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 3119, in process_stale_scc
    if not graph[id].type_check_second_pass():
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 2181, in type_check_second_pass
    return self.type_checker().check_second_pass()
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 344, in check_second_pass
    self.check_partial(node)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 356, in check_partial
    self.accept(node)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 401, in accept
    stmt.accept(self)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/nodes.py", line 687, in accept
    return visitor.visit_func_def(self)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 726, in visit_func_def
    self._visit_func_def(defn)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 737, in _visit_func_def
    self.check_method_override(defn)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 1420, in check_method_override
    if self.check_method_or_accessor_override_for_base(defn, base):
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 1448, in check_method_or_accessor_override_for_base
    if self.check_method_override_for_base_with_name(defn, name, base):
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 1509, in check_method_override_for_base_with_name
    assert False, str(base_attr.node)
AssertionError: Var(__hash__)
mypy_crash.py:5: : note: use --pdb to drop into pdb
```

**To Reproduce**

Contents of mypy_crash.py:
```python
class Base:
    __hash__ = None

class Dervied(Base):
    def __hash__(self) -> int:
        return 0
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: master (0.820+dev.15bd486a67a3214fdb854ba2f1d6d524113da6c1)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.6.12 (compiled from source with pyenv)
- Operating system and version: Ubuntu 20.04

<!--
You can freely edit this text, please remove all the lines
you believe are unnecessary.
-->

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1539,9 +1539,13 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 elif isinstance(original_node, Var):
                     # Super type can define method as an attribute.
                     # See https://github.com/python/mypy/issues/10134
+
                     # We also check that sometimes `original_node.type` is None.
                     # This is the case when we use something like `__hash__ = None`.
-                    original_type = original_node.type or NoneType()
+                    if original_node.type is not None:
+                        original_type = get_proper_type(original_node.type)
+                    else:
+                        original_type = NoneType()
                 else:
                     assert False, str(base_attr.node)
             if isinstance(original_node, (FuncDef, OverloadedFuncDef)):

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1536,6 +1536,12 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                     original_type = self.function_type(original_node)
                 elif isinstance(original_node, Decorator):
                     original_type = self.function_type(original_node.func)
+                elif isinstance(original_node, Var):
+                    # Super type can define method as an attribute.
+                    # See https://github.com/python/mypy/issues/10134
+                    # We also check that sometimes `original_node.type` is None.
+                    # This is the case when we use something like `__hash__ = None`.
+                    original_type = original_node.type or NoneType()
                 else:
                     assert False, str(base_attr.node)
             if isinstance(original_node, (FuncDef, OverloadedFuncDef)):

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -114,6 +114,43 @@ class A:
 A().f = None # E: Cannot assign to a method
 
 
+[case testOverrideAttributeWithMethod]
+# This was crashing:
+# https://github.com/python/mypy/issues/10134
+from typing import Protocol
+
+class Base:
+    __hash__ = None
+
+class Dervied(Base):
+    def __hash__(self) -> int:  # E: Signature of "__hash__" incompatible with supertype "Base"
+        pass
+
+# Correct:
+
+class CallableProtocol(Protocol):
+    def __call__(self, arg: int) -> int:
+        pass
+
+class CorrectBase:
+    attr: CallableProtocol
+
+class CorrectDerived(CorrectBase):
+    def attr(self, arg: int) -> int:
+        pass
+
+[case testOverrideMethodWithAttribute]
+# The reserse should not crash as well:
+from typing import Callable
+
+class Base:
+    def __hash__(self) -> int:
+        pass
+
+class Dervied(Base):
+    __hash__ = 1  # E: Incompatible types in assignment (expression has type "int", base class "Base" defined the type as "Callable[[Base], int]")
+
+
 -- Attributes
 -- ----------
 

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -122,7 +122,7 @@ from typing import Protocol
 class Base:
     __hash__ = None
 
-class Dervied(Base):
+class Derived(Base):
     def __hash__(self) -> int:  # E: Signature of "__hash__" incompatible with supertype "Base"
         pass
 

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -140,7 +140,7 @@ class CorrectDerived(CorrectBase):
         pass
 
 [case testOverrideMethodWithAttribute]
-# The reserse should not crash as well:
+# The reverse should not crash as well:
 from typing import Callable
 
 class Base:

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -147,7 +147,7 @@ class Base:
     def __hash__(self) -> int:
         pass
 
-class Dervied(Base):
+class Derived(Base):
     __hash__ = 1  # E: Incompatible types in assignment (expression has type "int", base class "Base" defined the type as "Callable[[Base], int]")
 
 

